neural_speed/convert/convert_quantized_llama.py

```diff
@@ -109,105 +117,232 @@ def main(args_in: Optional[List[str]] = None) -> None:
+        ffn_hidden_size = config["intermediate_size"]
+
+        # 1. write head and params
+        f.write(b"ggjt"[::-1])  # magic
+        rope_scale = 1
+        if "rope_scaling" in config and config["rope_scaling"] is not None:
+            rope_scale = config["rope_scaling"]["factor"] if "factor" in config["rope_scaling"] else 1
+        n_head = n_head
+        n_head_kv = config["num_key_value_heads"] if "num_key_value_heads" in config else n_head
+        values = [
+            1,  # file version
+            n_vocab,
+            n_embd,
+            256,  #hparams.n_mult,
+            n_head,
+            n_head_kv,  # n_head_kv (multi_query attention)
+            n_layer,
+            n_embd // n_head,  # rot (obsolete)
+            0,  #file_type.value, # TODO
+        ]
+        f.write(struct.pack("i" * len(values), *values))
+        f.write(struct.pack("i", 0))
+        f.write(struct.pack("f", 0))
+        f.write(struct.pack("f", 0))
+        f.write(struct.pack("i", 0))
+        f.write(struct.pack("i", 0))  # word_embed_proj_dim (for opt)
+        f.write(struct.pack("i", 0))  # do_layer_norm_before (for opt)
+
+        f.write(struct.pack("i", 0))
+        f.write(struct.pack("i", ffn_hidden_size))
+        f.write(struct.pack("i", 0))
+        f.write(struct.pack("i", 0))  # n_experts
+        f.write(struct.pack("i", 0))  # n_expert_used
+        f.write(struct.pack("i", 0))  # n_embd_head_k for gemma
+
+        f.write(struct.pack("f", config["rms_norm_eps"]))
+        f.write(struct.pack("f", config["rope_theta"] if "rope_theta" in config else 10000))
+        f.write(struct.pack("f", rope_scale))
+
+        f.write(struct.pack("f", 0.0))  # config.json "rope_scaling.factor", not enabled
+        f.write(struct.pack("i", 0))  # rope_scaling.original_max_position_embeddings
+        f.write(struct.pack("i", 0))  # params["rope_scaling"]["type"] =="yarn" else 0))
+
+        # TODO, bos_token_id = 0 in https://huggingface.co/decapoda-research/llama-7b-hf/blob/main/config.json
+        # but bos_token_id = 1 in llama.cpp
+        f.write(struct.pack("i", 1))
+        f.write(struct.pack("i", 2))
+
+        f.write(struct.pack("i", 0))
+        f.write(struct.pack("i", 0))
+
+        # 2. vocab
+        tokenizer_path = os.path.join(model_path, "")
+        vocab = BpeVocab(Path(tokenizer_path))
+
+        for text, score in vocab.all_tokens():
+            if isinstance(text, str):
+                text = text.encode("utf-8")
+            f.write(struct.pack("i", len(text)))
+            f.write(text)
+            f.write(struct.pack("f", score))
+
+        # 3. write tensors
+        list_vars = model
+        convert_to_fp32_tensor("model.embed_tokens.weight", "tok_embeddings.weight", list_vars, f)
+        convert_to_fp32_tensor("model.norm.weight", "norm.weight", list_vars, f)
+        convert_to_q4_bestla_tensor(f"lm_head.weight", f"output.weight", list_vars, f, quantize_config, n_head)
```

Comment:
what if we get a lm_head.weight dtype = fp16 llama3 model ?

Reply:
Not sure whether there is a validated FP16->int4 conversion pipeline in the Neural Speed, but it shouldn't have too many problems with a little effort.

At least we can convert back fp16 -> fp32 -> int4 and do the same pipeline we already have.

Reply:
Not sure whether there is a validated conversion pipeline, but should be ok.